### PR TITLE
[#74] Using Bulk artifact resolve operations

### LINKED ISSUE #74: Support bulk artifact resolution to speed up downloads
Aether's RepositorySystem supports bulk Maven operations. When resolving large amounts of artifacts those are more efficient then individual operations.

The proposal is to add following methods to the Channels API:
- `Channel#resolveArtifacts(List<? extends ArtifactCoordinate>)`
- `ChannelSession#resolveDirectMavenArtifacts(List<? extends ArtifactCoordinate>)`
- `ChannelSession#resolveMavenArtifacts(List<? extends ArtifactCoordinate>)`

Related to https://issues.redhat.com/browse/GAL-333 and https://issues.redhat.com/browse/WFGP-230

### PR DESCRIPTION
**Pull request by spyrkob**
Fixes #74

### CODE REVIEW

File: core/src/main/java/org/wildfly/channel/DefaultArtifactCoordinate.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * limitations under the License.
+ */
+
+package org.wildfly.channel;
+
+import java.util.Objects;
+
+import static java.util.Objects.requireNonNull;
+
+public class DefaultArtifactCoordinate implements ArtifactCoordinate {
```
Comment (jmesnil) [inline core/src/main/java/org/wildfly/channel/DefaultArtifactCoordinate.java:24]: We could make the ArtifactCoordinate a class and merge the `ArtifactCoordinate` and `DefaultArtifactCoordinate` together. wdyt?

File: core/src/main/java/org/wildfly/channel/DefaultArtifactCoordinate.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+   protected final String groupId;
+   protected final String artifactId;
+   protected final String extension;
+   protected final String classifier;
+   protected final String version;
+
+   public DefaultArtifactCoordinate(String groupId, String artifactId, String extension, String classifier, String version) {
+      requireNonNull(groupId);
+      requireNonNull(artifactId);
+
```
Comment (jmesnil) [inline core/src/main/java/org/wildfly/channel/DefaultArtifactCoordinate.java:35]: it's missing a `requireNonNumm(version)`

File: maven-resolver/src/main/java/org/wildfly/channel/maven/VersionResolverFactory.java [hunk 3]
```diff
@@ -125,6 +127,38 @@ public File resolveArtifact(String groupId, String artifactId, String extension,
...
             return result.getArtifact().getFile();
         }
+
+        @Override
+        public List<File> resolveArtifacts(List<? extends ArtifactCoordinate> coordinates) throws UnresolvedMavenArtifactException {
+            requireNonNull(coordinates);
+
+            List<ArtifactRequest> requests = new ArrayList<>();
+            for (ArtifactCoordinate coord : coordinates) {
+                requireNonNull(coord.getGroupId());
```
Comment (jmesnil) [inline maven-resolver/src/main/java/org/wildfly/channel/maven/VersionResolverFactory.java:137]: you should not need it as the `ArtifactCoordinate` ctor wil already have these `requireNonNull`

File: core/src/main/java/org/wildfly/channel/spi/MavenVersionsResolver.java [hunk 4]
```diff
@@ -55,6 +58,19 @@ public interface MavenVersionsResolver extends Closeable {
...
+   /**
+    * Resolve a list of maven artifacts based on the full coordinates.
+    *
+    * The order of returned Files is the same as order of coordinates.
+    *
+    * @param coordinates - list of ArtifactCoordinates. They need contain at least groupId, artifactId and version.
+    *
+    * @return a list of File representing the resolved Maven artifact.
+    *
+    * @throws UnresolvedMavenArtifactException if th artifact can not be resolved.
```
Comment (jmesnil) [inline core/src/main/java/org/wildfly/channel/spi/MavenVersionsResolver.java:70]: should be?

`@throws UnresolvedMavenArtifactException if any artifact can not be resolved.`

**Review by jmesnil — CHANGES_REQUESTED** [review]
@spyrkob I like this idea.

File: core/src/main/java/org/wildfly/channel/ChannelRecorder.java [hunk 5]
```diff
@@ -16,23 +16,46 @@
...
     void recordStream(String groupId, String artifactId, String version) {
-        boolean isRecorded = recordedChannel.getStreams().stream().anyMatch(s -> s.getGroupId().equals(groupId) && s.getArtifactId().equals(artifactId) && s.getVersion().equals(version));
-        if (!isRecorded) {
-            recordedChannel.addStream(new Stream(groupId, artifactId, version, null));
-        }
+        streams.putIfAbsent(groupId + ":" + artifactId + ":" + version, new Stream(groupId, artifactId, version, null));
+    }
+
+    Channel getRecordedChannel() {
+        List<Stream> sortedStreams = new ArrayList<>(streams.values());
```
[the remark this answers is not in the document]
Comment (spyrkob) [reply core/src/main/java/org/wildfly/channel/ChannelRecorder.java:37]: I missed that change, sorry. I'll remove it.

**Comment by spyrkob** [comment]
Pushed the changes.

Thinking a bit more about the ArtifactCoordinate, I'm wondering if it'd be worth to change it to:
```
ArtifactCoordinate (doesn't enforce anything)
 | - StreamArtifactCoordinate (requires only groupId and artifactId)
 | - ResolvabelArtifactCoordinate(requires groupId, artifactId and version)
 |   |- MavenArtifact(requires GAV and file)
 | - ChannelCoordinate (requires either GA, GAV or file)
```

This way eg. MavenVersionResolver#resolveArtifact can rely on ResolvabelArtifactCoordinate and ChannelSession#resolveMavenArtifacts can use StreamArtifactCoordinate to pass incomplete coordinate.

**Comment by jmesnil** [comment]
I think we can simplify the coordinate tree. 
We could separate the `ChannelCoordinate` from the rest as it is used in a different code path.

```
ChannelCoordinate (requires either GA, GAV or file)
```

Then for the streams coordinate, we need 2 levels:

```
class ArtifactCoordinate (requires non-null groupId, artifactId, version)
   -> all API method requires to pass the base version of the coordinate when using the API
  |- class MavenArtfifact (require non-null file)
```

**Comment by spyrkob** [comment]
> all API method requires to pass the base version of the coordinate when using the API
I forgot there's the base version as well. 

Separated ChannelCoordinate and made version required in ArtifactCoordinate

**Review by jmesnil — APPROVED** [review]

**Comment by jmesnil** [comment]
@spyrkob thanks!

**Pull request merged**